### ISSUE TITLE
Import errors when using cx_Freeze with specific scipy modules

### ISSUE THREAD

**Issue by anthony-tuininga** [marcelotduarte/cx_Freeze#43]
Originally reported by: **azylstra (Bitbucket: [azylstra](https://bitbucket.org/azylstra), GitHub: [azylstra](https://github.com/azylstra))**

----------------------------------------

Following up on [StackOverflow issue](http://stackoverflow.com/questions/19594567/importerror-when-using-cx-freeze-with-scipy)

My project uses scipy in a few places and after freezing I would get a variety of import errors. Those seemed to be naming issues (see in the StackOverflow link, the executable was looking for `_csr` when `scipy.sparse.sparsetools._csr` was what was packaged).

Unfortunately the project cannot be public but I tried to replicate the issue in an example. The `_csr` problem appears to be coming from a piece of code that uses `import scipy.optimize`. Other modules when imported specifically cause other errors. For example, here is some logged console output from an executable:

    10/27/13 9:58:08.856 AM Python[8116]: Start cx_Freeze test
    10/27/13 9:58:08.959 AM Python[8116]: Imported scipy OK
    10/27/13 9:58:08.961 AM Python[8116]: Uh oh, failed to import scipy.interpolate: No module named 'scipy.special._ufuncs_cxx'
    10/27/13 9:58:08.962 AM Python[8116]: Uh oh, failed to import scipy.integrate: No module named 'scipy.special._ufuncs_cxx'
    10/27/13 9:58:09.020 AM Python[8116]: Uh oh, failed to import scipy.optimize: No module named '_csr'
    10/27/13 9:58:09.023 AM Python[8116]: Uh oh, failed to import scipy.stats: No module named 'scipy.special._ufuncs_cxx'
    10/27/13 9:58:09.023 AM Python[8116]: cx_Freeze test completed

I've attached a zip of the setup.py and python code used to generate that output, along with a text file containing the console output of the build, and the entire build folder.

Here are the details of my system:

* OS X 10.9
* python 3.3.2
* cx_Freeze 4.3.2
* scipy 0.13.0

----------------------------------------
- Bitbucket: https://bitbucket.org/anthony_tuininga/cx_freeze/issue/43

**Comment by anthony-tuininga** [comment]
*Original comment by* **azylstra (Bitbucket: [azylstra](https://bitbucket.org/azylstra), GitHub: [azylstra](https://github.com/azylstra))**:

----------------------------------------

FYI I wrote the following rules as a work-around to get scipy.interpolate to work:

	"include_files": [('/usr/local/lib/python3.3/site-packages/scipy/sparse/sparsetools/_csr.so','_csr.so'),
	('/usr/local/lib/python3.3/site-packages/scipy/sparse/sparsetools/_csc.so','_csc.so'),
	('/usr/local/lib/python3.3/site-packages/scipy/sparse/sparsetools/_coo.so','_coo.so'),
	('/usr/local/lib/python3.3/site-packages/scipy/sparse/sparsetools/_dia.so','_dia.so'),
	('/usr/local/lib/python3.3/site-packages/scipy/sparse/sparsetools/_bsr.so','_bsr.so'),
	('/usr/local/lib/python3.3/site-packages/scipy/sparse/sparsetools/_csgraph.so','_csgraph.so')]

**Comment by anthony-tuininga** [comment]
*Original comment by* **Martin Fitzpatrick (Bitbucket: [mfitzp](https://bitbucket.org/mfitzp), GitHub: [mfitzp](https://github.com/mfitzp))**:

----------------------------------------

Just to add I'm also having a similar problem on Windows. Using anything that imports `scipy.optimize` results in it breaking at `from scipy.optimize import minpack2` with `ImportError: DLL load failed: The specified module could not be found`.

Strangely though if I look in the library.zip there is a `minpack2.pyc` file.

(Later...)

Based on what azylstra wrote above I realised it's not the `.pyc` files it's looking for but the `.pyd`s (the Windows equivalent of `.so`). So a simple hack to get around this problem for the particular modules experience them was:

        import os, glob2, numpy, scipy, nmrglue
        explore_dirs = [
            os.path.dirname(numpy.__file__),
            os.path.dirname(scipy.__file__),
            os.path.dirname(nmrglue.__file__),
        ]
        
        files = []
        for d in explore_dirs:
            files.extend( glob2.glob( os.path.join(d, '**', '*.pyd') ) )
            
        # Now we have a list of .pyd files; iterate to build a list of tuples into 
        # include files containing the source path and the basename
        for f in files:
            build_all['include_files'].append( (f, os.path.basename(f) ) )

This is an ugly hack - copying all `.pyd` files from each module regardless of whether they are used - and results in bigger packages. But it works - thought it may be of use to others.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

@mfitzp : the .pyc file is a wrapper to load the extension module - it should be copied into the build directory (outside the zip file) as something like `scipy.optimize.minpack2.pyd` - is that there? Can you show the log from freezing?

I think the "DLL load failed" message means it's trying to dynamically load another DLL, which may not have been copied. You might be able to use process explorer on a Python process with that module loaded to see what else it loads.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Ian Bell (Bitbucket: [ibell](https://bitbucket.org/ibell), GitHub: [ibell](https://github.com/ibell))**:

----------------------------------------

+1 on this issue.  

I have heaps of problems with scipy and cx_Freeze.  I get the same error with 

scipy.special._ufuncs_cxx

and

scipy.sparse.csgraph._validation

not being found.  It seems there is something wrong with the search for all things that are part of a package.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

I think the trouble is actually with SWIG. It does some clever things to try to do relative imports in a backwards compatible way - see this function in Scipy:
https://github.com/scipy/scipy/blob/master/scipy/sparse/sparsetools/csr.py#L11

That function won't work in a frozen application. The critcal part is the fallback on line 18 - if I change that to `from . import _csr` it works (until the next such import - that autogenerated function exists in several places).

The code is generated [here](https://github.com/swig/swig/blob/fb96ff9613215782e65ff8500354b518e9d5cf06/Source/Modules/python.cxx#L825) in SWIG.

I'm filing an issue on SWIG to bring this to their attention.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

https://github.com/swig/swig/issues/145

**Comment by anthony-tuininga** [comment]
*Original comment by* **Ian Bell (Bitbucket: [ibell](https://bitbucket.org/ibell), GitHub: [ibell](https://github.com/ibell))**:

----------------------------------------

Great - thanks for the feedback.  Would be great to get this working.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Marcin (Bitbucket: [bosmart](https://bitbucket.org/bosmart), GitHub: [bosmart](https://github.com/bosmart))**:

----------------------------------------

Facing the same problem, so +1.

Configuration:
- Windows 7 x64
- Python 2.7.6 x64
- SciPy 0.13 x64
- cx_Freeze 4.3.2

Script (example from scipy's documentation):

```
#!python

import numpy as np
from scipy.interpolate import griddata


def func(x, y):
    return x * (1 - x) * np.cos(4 * np.pi * x) * np.sin(4 * np.pi * y ** 2) ** 2


grid_x, grid_y = np.mgrid[0:1:100j, 0:1:200j]
points = np.random.rand(1000, 2)
values = func(points[:, 0], points[:, 1])
grid_z0 = griddata(points, values, (grid_x, grid_y), method='nearest')
```


Build command:
C:\python27\scripts\cxfreeze cxTest.py  --target-dir C:\temp\dist

Error:
Traceback (most recent call last):
  File "C:\Python27\lib\site-packages\cx_Freeze\initscripts\Console.py", line 27, in <module>
    exec code in m.__dict__
  File "cxTest.py", line 2, in <module>
  File "C:\Python27\lib\site-packages\scipy\interpolate\__init__.py", line 156, in <module>
    from .interpolate import *
  File "C:\Python27\lib\site-packages\scipy\interpolate\interpolate.py", line 12, in <module>
    import scipy.special as spec
  File "C:\Python27\lib\site-packages\scipy\special\__init__.py", line 531, in <module>
    from ._ufuncs import *
  File "ExtensionLoader_scipy_special__ufuncs.py", line 22, in <module>
  File "ExtensionLoader_scipy_special__ufuncs.py", line 14, in __bootstrap__
ImportError: DLL load failed: The specified module could not be found.

**Comment by anthony-tuininga** [comment]
*Original comment by* **BubuIIC (Bitbucket: [BubuIIC](https://bitbucket.org/BubuIIC), GitHub: Unknown)**:

----------------------------------------

Another report here. I'm trying to do the following:

```
#!python
from scipy import linalg
```
And get the following error:

```
#!python

  from scipy import linalg
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1616, in _handle _fromlist
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 313, in _call_with_frames_removed
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1567, in _find_and_load
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1534, in _find_and_load_unlocked
  File "C:\Python33\lib\site-packages\scipy\linalg\__init__.py", line 158, in <module>
    from .basic import *
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1567, in _find_and_load
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1534, in _find_and_load_unlocked
  File "C:\Python33\lib\site-packages\scipy\linalg\basic.py", line 17, in <module>
    from scipy.linalg import calc_lwork
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1616, in _handle_fromlist
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 313, in _call_with_frames_removed
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1567, in _find_and_load
  File "C:\Python\64-bit\3.3\lib\importlib\_bootstrap.py", line 1534, in _find_and_load_unlocked
  File "ExtensionLoader_scipy_linalg_calc_lwork.py", line 22, in <module>
  File "ExtensionLoader_scipy_linalg_calc_lwork.py", line 14, in __bootstrap__
ImportError: DLL load failed: The specified module could not be found.
```


The scipy.linalg.calc_lwork.pyd exists in the build directory.

Is there any workaround or fix for this?

**
Edit:** Ok, I figured out which dll's were missing. turns out these actually belong to numpy and not Scipy. The .pyd file listed above needs libifcoremd.dll which then needs libmmd.dll. If I copy these two from the numpy install to the build dir, the frozen app works. (Found this out using http://www.dependencywalker.com/).

So responsible for this is still scipy which is doing strange imports? Or can this be fixed in cx_freeze? How can I add the .dll's to the build process from cx_freeze without specifying the absolute path?

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

That last one sounds like a different problem - it appears cx_Freeze isn't finding some DLLs properly. They might be loaded dynamically, in which case a hook will be needed.

**Comment by anthony-tuininga** [comment]
*Original comment by* **BubuIIC (Bitbucket: [BubuIIC](https://bitbucket.org/BubuIIC), GitHub: Unknown)**:

----------------------------------------

I have looked into cx_freeze hooks, but I have no idea on what I should hook (I guess on scipy.linalg, but that doesn't seem to work). And then, how I can point cx_freeze to the right .dlls.

Any more hints on this?

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

A `load_scipy_linalg` hook should be right. If the DLLs are importable Python modules, you should be able to add them with `finder.IncludeModule()` in the hook. Otherwise, you'll need to find where they're located and use `finder.IncludeFiles()`.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Marcin (Bitbucket: [bosmart](https://bitbucket.org/bosmart), GitHub: [bosmart](https://github.com/bosmart))**:

----------------------------------------

I'm confused now... so I can get this to work manually somehow, right? How? I'm on Windows with Python 2.7.6 x64 and I'm getting this:
![Untitled.png](https://bitbucket.org/repo/oE5Lk/images/3625045746-Untitled.png)

Help!

**Comment by anthony-tuininga** [comment]
*Original comment by* **Ian Bell (Bitbucket: [ibell](https://bitbucket.org/ibell), GitHub: [ibell](https://github.com/ibell))**:

----------------------------------------

It's ugly as hell, but you can just copy all PYD files from scipy, see my
script for PDSim : https://github.com/ibell/pdsim/blob/master/GUI/setup.py

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

Can someone try with the beta version of scipy 0.14? They seem to be moving away from SWIG, which I think will fix the issue, though I don't know if they've moved entirely away from it yet.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Marcin (Bitbucket: [bosmart](https://bitbucket.org/bosmart), GitHub: [bosmart](https://github.com/bosmart))**:

----------------------------------------

Copying all PYD, PYC, PYO and DLL files didn't help, although the error message is now different - "missing _validate".

Will try scipy 0.14 later and report on the result.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Ian Bell (Bitbucket: [ibell](https://bitbucket.org/ibell), GitHub: [ibell](https://github.com/ibell))**:

----------------------------------------

I take it you mean _validation, if you notice my script, I added that to
includes :

includes = ['numpy','scipy.sparse.csgraph._validation']

**Comment by anthony-tuininga** [comment]
*Original comment by* **Marcin (Bitbucket: [bosmart](https://bitbucket.org/bosmart), GitHub: [bosmart](https://github.com/bosmart))**:

----------------------------------------

Yep, validate sorted out. with scipy 0.14 I don't need to copy any DLLs, yet I still get "importerror no module named _ufuncs_cxx"...

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

Gah, it's one extension module trying to load another extension module. I don't know quite what to do about that.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Marcin (Bitbucket: [bosmart](https://bitbucket.org/bosmart), GitHub: [bosmart](https://github.com/bosmart))**:

----------------------------------------

Ok, got it!

Two things I had to do:
1. copy all the numpy DLLs (from c:\Python27\Lib\site-packages\numpy\core), regardless of using scipy 13 or 14
2. add the following to my code:

```
#!python
def dependencies_for_myprogram():
    from scipy.sparse.csgraph import _validation
    from scipy.special import _ufuncs_cxx
    from matplotlib.backends import backend_tkagg #this one is for matplotlib
```
Now I have a new problem with multiprocessing, which opens multiple windows of my app...

UPDATE: multiprocessing issue sorted out by:

```
#!python

if __name__ == '__main__':
    # freeze_support must be the first line
    multiprocessing.freeze_support()
```

:)

**Comment by anthony-tuininga** [comment]
*Original comment by* **Lucas Hermann Negri (Bitbucket: [lucashnegri](https://bitbucket.org/lucashnegri), GitHub: Unknown)**:

----------------------------------------

I've fixed it (problem with _ufuncs) by coping libifcoremd.dll and libmmd.dll from the numpy\core directory.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Adam H (Bitbucket: [hughesadam87](https://bitbucket.org/hughesadam87), GitHub: Unknown)**:

----------------------------------------

Where do you copy them to exactly?  I'm getting this problem as well.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

Probably to the build directory, next to your exe.

My approach now to solving this, and many other issues, is not to store packages inside a zip file by default (see pull request #73). If packages are loaded from regular files and folders, this kind of issue shouldn't come up. You'll still be able to tell it to zip up packages you know work that way.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Adam H (Bitbucket: [hughesadam87](https://bitbucket.org/hughesadam87), GitHub: Unknown)**:

----------------------------------------

In current version of cx_freeze, at least for my uses, just putting "scipy" into the packages got rid of this issue...

**Comment by anthony-tuininga** [comment]
*Original comment by* **Thomas Kluyver (Bitbucket: [takluyver](https://bitbucket.org/takluyver), GitHub: [takluyver](https://github.com/takluyver))**:

----------------------------------------

They'll only be DLLs on Windows. On Linux, the equivalent files are .so, and on Mac they're .dylib, I think.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Florian Zierer (Bitbucket: [ziererf](https://bitbucket.org/ziererf), GitHub: Unknown)**:

----------------------------------------

Just want to tell you, that the new scipy 0.16.0 (I installed the windows wheels from http://www.lfd.uci.edu/~gohlke/pythonlibs/) does no longer work when freezing them.

For 0.15.1 I could just add "scipy" to the packages option (as @hughesadam87 said) and all is working, but not for 0.16.0

**Comment by anthony-tuininga** [comment]
*Original comment by* **Franco C (Bitbucket: [francoc](https://bitbucket.org/francoc), GitHub: [francoc](https://github.com/francoc))**:

----------------------------------------

same problem here, updating to scipy 0.16.0 from gohlke pythonlibs prevents cx_freeze to build exe. 
If "scipy" is added to packages options this error is raised:
ImportError: No module named 'scipy'

If no scipy is added to packages options, build runs until end but upon executing .exe an error window is raised with:
ImportError: No module named 'scipy.__config__'

**Comment by anthony-tuininga** [comment]
*Original comment by* **Keegan Owsley (Bitbucket: [keegan_owsley](https://bitbucket.org/keegan_owsley), GitHub: Unknown)**:

----------------------------------------

To get latest scipy 0.16 package from gohlke to freeze, I had to change line 548 of hooks.py to read:

    finder.IncludePackage("scipy._lib")

Note the extra underscore.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Adam H (Bitbucket: [hughesadam87](https://bitbucket.org/hughesadam87), GitHub: Unknown)**:

----------------------------------------

Did this freeze the entire scipy library?

**Comment by anthony-tuininga** [comment]
*Original comment by* **Keegan Owsley (Bitbucket: [keegan_owsley](https://bitbucket.org/keegan_owsley), GitHub: Unknown)**:

----------------------------------------

It appears so. I still need to include the libmmd and libifcoremd binaries as mentioned elsewhere in the thread, and list "scipy" manually in the packages as a build option.

Edit: I am only using a few submodules, so I don't know if (eg) scipy.sparse.csgraph._validation works or not. I also have not tried just omitting the line entirely.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Florian Zierer (Bitbucket: [ziererf](https://bitbucket.org/ziererf), GitHub: Unknown)**:

----------------------------------------

Ok, so I was updating to scipy 0.16.1 and encountered some combination of the above posted errors.

I had to change the hooks.py as @keegan_owsley wrote and to include scipy as a package.
Additionally I also had the problem that the two numpy dlls (libifcoremd.dll and libmmd.dll) where not included.
It helped to *import numpy* at the beginning of the setup.py

**Comment by anthony-tuininga** [comment]
*Original comment by* **Jürg Rast (Bitbucket: [jrast](https://bitbucket.org/jrast), GitHub: [jrast](https://github.com/jrast))**:

----------------------------------------

I expirienced the same problems and fixed it by patching the hooks function and including numpy on top of the setup.py. I think patching the hook function in setup.py is nicer than editing the code directly:


```
#!python

from cx_Freeze import hooks

def load_scipy_patched(finder, module):
    """the scipy module loads items within itself in a way that causes
        problems without the entire package and a number of other subpackages
        being present."""
    finder.IncludePackage("scipy._lib")  # Changed include from scipy.lib to scipy._lib
    finder.IncludePackage("scipy.misc")

hooks.load_scipy = load_scipy_patched
```

**Comment by anthony-tuininga** [comment]
*Original comment by* **Anthony Tuininga (Bitbucket: [anthony_tuininga](https://bitbucket.org/anthony_tuininga), GitHub: Unknown)**:

----------------------------------------

This issue should be resolved with cx_Freeze 5.0. Please re-open if this is not the case. Thanks.

**Comment by anthony-tuininga** [comment]
*Original comment by* **Rui Da Costa (Bitbucket: [ruidc](https://bitbucket.org/ruidc), GitHub: [ruidc](https://github.com/ruidc))**:

----------------------------------------

It doesn't appear fixed yet in 5.0.1, I had to make the same patch.
